src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -157,39 +162,182 @@ internal void ClearCaches()
+
+                Debug.Assert(!entry.HasResult);
+                CacheEntry? candidate = null;
+                Type? candidateType = null;
+
+                for (Type? current = type.BaseType; current != null; current = current.BaseType)
+                {
+                    if (current == JsonTypeInfo.ObjectType)
+                    {
+                        // Avoid falling back to the contract for object since it's polymorphic
+                        // and it would try to send us back to the runtime type that isn't supported.
+                        break;
+                    }
+
+                    candidate = GetOrAddCacheEntry(current);
+                    if (candidate.HasResult)
+                    {
+                        // We found a type in the class hierarchy that has a contract -- stop looking further up.
+                        candidateType = current;
+                        break;
+                    }
+                }
+
+                foreach (Type interfaceType in type.GetInterfaces())
+                {
+                    CacheEntry interfaceEntry = GetOrAddCacheEntry(interfaceType);
+                    if (interfaceEntry.HasResult)
+                    {
+                        if (candidateType != null)
+                        {
+                            if (interfaceType.IsAssignableFrom(candidateType))
+                            {
+                                // The previous candidate is more derived than the
+                                // current interface -- keep our previous choice.
+                                continue;
+                            }
+                            else if (candidateType.IsAssignableFrom(interfaceType))
+                            {
+                                // The current interface is more derived than the
+                                // previous candidate -- replace the candidate value.
+                            }
+                            else
+                            {
+                                // We have found two possible ancestors that are not in subtype relationship.
+                                // This indicates we have encountered a diamond ambiguity -- abort search and record an exception.
+                                NotSupportedException nse = ThrowHelper.GetNotSupportedException_AmbiguousMetadataForType(type, candidateType, interfaceType);
+                                candidate = new CacheEntry(ExceptionDispatchInfo.Capture(nse));
+                                break;
+                            }
+                        }
+
+                        candidate = interfaceEntry;
+                        candidateType = interfaceType;
+                    }
+                }
+
+                entry.NearestAncestor = candidate;
+                entry.IsNearestAncestorResolved = true;
+                return candidate;
+            }
+
+            private sealed class CacheEntry
```

Comment:
A class is needed because nearest-ancestor resolution is done lazily on the created `CacheEntry`. This is because the algorithm needs to walk the entire hierarchy, necessitating multiple cache lookups that can't be inserted into a single `ConcurrentDictionary.GetOrAdd` operation. But because you need to mutate the entry after its insertion, it has to be a class instead of, say, a struct.